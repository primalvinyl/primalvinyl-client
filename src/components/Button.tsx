import React from 'react';

const Button = React.forwardRef<HTMLButtonElement, ButtonType>(
    (props, ref): React.ReactElement => {
    const {
        id,
        value,
        type,
        isSubmitting,
        handleClick
    } = props;

    return (
        <div className="form-group">
            <button
                className="form-control"
                id={id}
                type={type}
                ref={ref}
                onClick={handleClick}
                disabled={isSubmitting}
            >
                {value}
            </button>
        </div>
    );
});

type ButtonType = {
    readonly id: string;
    readonly value: string;
    readonly type: 'submit' | 'button' | 'reset';
    readonly isSubmitting: boolean;
    readonly handleClick: (event: React.MouseEvent) => void;
};

Button.defaultProps = {
    id: '',
    value: '',
    type: 'submit',
    isSubmitting: false,
    handleClick: (): void => {},
}

export default Button;
