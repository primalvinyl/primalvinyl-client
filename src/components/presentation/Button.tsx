import React from 'react';
import styled from 'styled-components';

//required syntax for styled component to work with typescript
const WrapperComponent: React.FunctionComponent<WrapperComponentProps> = 
({ className, children }) => (
    <div className={className}>{children}</div>
);
interface WrapperComponentProps {
    readonly className?: string;
    readonly background?: string;
    readonly color?: string;
};
const Wrapper = styled(WrapperComponent)`
    padding: 0;
    margin: 0;
    width: 100%;

    //resets begin
    button {
        font-size: 100%;
        margin: 0;
        vertical-align: baseline;
        *vertical-align: middle;
        line-height: normal;
        text-transform: none;
        -webkit-appearance: button;
        cursor: pointer;
        *overflow: visible;
        color: #222;
    }

    button::-moz-focus-inner {
        border: 0;
        padding: 0;
    }
    //resets end

    button {
        color: ${props => props.color};
        background: ${props => props.background};
        font-size: 14px;
        font-weight: bold;
        text-transform: uppercase;
        width: 100%;
        padding: .7em;
        border: none;
        border-radius: 5px; 

        &[disabled] {
            cursor: default;
            color: gray;
        }
    }
`;

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
    (props, ref): React.ReactElement => {

        const {
            id,
            className,
            name,
            children,
            type,
            disabled,
            isSubmitting,
            onClick,
            background,
            color
        } = props;

        return (
            <Wrapper background={background} color={color}> 
                <button
                    className={className}
                    id={id}
                    name={name}
                    type={type}
                    disabled={isSubmitting || disabled}
                    ref={ref}
                    onClick={(event) => onClick && onClick(event)}
                >
                    {children}
                </button>
            </Wrapper>
        );
    });

type ButtonProps = {
    readonly id?: string;
    readonly className?: string;
    readonly name?: string;
    readonly children?: any;
    readonly type?: 'submit' | 'button' | 'reset';
    readonly disabled?: boolean;
    readonly isSubmitting?: boolean;
    readonly background?: string;
    readonly color?: string;
    readonly onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
};

Button.defaultProps = {
    id: undefined,
    className: undefined,
    name: undefined,
    type: 'submit',
    disabled: false,
    isSubmitting: false,
    background: 'none',
    color: '#222',
    onClick: () => {}
}

export default Button;
