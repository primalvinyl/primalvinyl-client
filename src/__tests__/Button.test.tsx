import * as React from 'react';
import { shallow } from 'enzyme';
import Renderer from 'react-test-renderer';
import Button from '../components/presentation/Button';

describe('Button', () => {
    it('renders', () => {
        const myWrapper = shallow(<Button />);
        expect(myWrapper.exists());
    });
    it('matches snapshot', () => {
        const component = Renderer.create(<Button />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});