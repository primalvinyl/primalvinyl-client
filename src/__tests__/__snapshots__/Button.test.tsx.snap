// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<div
  className="sc-bdfBwQ dSKHTO"
>
  <button
    disabled={false}
    id="test"
    onClick={[Function]}
    type="submit"
  />
</div>
`;
